docs/RFCS/stateless_replica_relocation.md

```diff
@@ -0,0 +1,232 @@
+the replication anomaly through the normal operation of the replica scanner.
+
+### Lack of non-voting replicas
+Some types of relocation are fragile without non-voting member replicas.
+
+When initially creating a replica, it is very far behind the current state of
+the range and thus needs to receive a snapshot. It may take some time before the
+range fully catches up and can take place in quorum commits.
+
+However, without non-voting replicas we have no choice but to add the new
+replica as a full member, thus changing the quorum requirements of the group. In
+the case of odd-numbered ranges, this will increase the quorum count by one,
+with the new range unable to be part of a quorum decision. This increases the
+chance of losing quorum until that replica is caught up, thus reducing
+availability.
+
+Stateless relocation will still work correctly when non-voting replicas are
+implemented; there will simply be a delay in the case where a replica is added
+first (e.g.  rebalance), with the removal not taking place until the non-voting
+replica catches up and is upgraded to a full group member.
+
+However, this is listed this as a drawback because going forward with relocation
+*without* non-voting replication introduces this fragility, regardless of how
+relocation decisions are made.
+
+# Alternatives
+The main alternative would be some sort of stateful system, where relocation
+operations would be expressed explicitly (i.e. "Move replica from X to Y") and
+then seen to completion. For reasons outlined in the "motiviation" section, this
+is considered sub-optimal when making distributed decisions.
+
+### Relocation Master
+The ultimate expression of this would be the designation of a "Relocation
+master", a single node that makes all relocation decisions for the entire
+cluster
+
+There is enough information in gossip for a central node to make acceptable
+decisions about replica movement.  In some ways the individual decisions would
+be worse, because they would be unable to consider current raft state; however,
+in aggregate the decisions could be much better, because the central master
+could consider groups of relocations together. For example, it would be able to
+avoid distributed pitfalls such as under-loaded nodes being suddenly inundated
+with requests for new replicas. It would be able to more quickly and correctly
+move replicas to new nodes introduced to the system.
+
+This system would also have an easier time communicating the individual
+relocation decisions that were made. This could be helpful for debugging or
+tweaking relocation criteria.
+
+However, It's important to note that a relocation master is not entirely
+incompatible with the "stateless" design; the relocation master could simply be
+the initiator of the stateless operations. You could thus get much of the
+improved decision making and communication of the master without having to move
+to a stateful design.
+
```

Comment:
I think we might eventually implement some kind of cluster master to keep track of frequent failures, etc. But I can see that master providing information for the individual nodes to make better decisions. The intelligence is stateful, not the mechanism.